### DISCUSSION TITLE
joint js recreate graph

### DISCUSSION THREAD

**Discussion by adarshjalaja** [clientIO/joint#2261] [Q&A]
### Introduction

is it possible to recreate graph other than using fromJSON API (saved graph recreate).

by defining each element graph (in an component)  and trying to redraw graph  based on ur custom JSON (business Logic JSON). how do we get graph equivalent of each element we introduced in canvas and try to keep it as static and redraw later 
as per our logic.

is there an example we can follow.

Thanks in Advance.

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS

**Comment by kumilingus** [comment]
The graph can be populated programmatically. Please refer to [Graph API](https://resources.jointjs.com/docs/jointjs/v3.7/joint.html#dia.Graph.constructor).

A simple custom JSON import/export example:
```js
// Import
const customJSON = [{ customType: 'rectangle' }, { customType: 'circle' }];
customJSON.forEach(type => {
    switch (type) {
        case 'rectangle': {
            const rect = new joint.shapes.standard.Rectangle();
            graph.addCells(rect);
            break;
        }
        case 'circle': {
            const circle = new joint.shapes.standard.Circle();
            graph.addCells(circle);
            break;
        }
    }
});

// Export
const customJSON2 = graph.getElements().map(cell => {
    const customType = {
        rectangle: 'standard.Rectangle',
        circle: 'standard.Circle',
    }[cell.get('type')];
    return { customType };
});

```
You can save additional information in the JSON such as `color`  and `text`.

You can save the position of each cell in JSON or you can call one of the automatic layouts once you add the cells to the graph.
- https://resources.jointjs.com/docs/jointjs/v3.7/joint.html#layout.DirectedGraph
- https://resources.jointjs.com/docs/rappid/v3.7/layout.ForceDirected.html
- https://resources.jointjs.com/docs/rappid/v3.7/layout.GridLayout.html
- https://resources.jointjs.com/docs/rappid/v3.7/layout.StackLayout.html
- https://resources.jointjs.com/docs/rappid/v3.7/layout.TreeLayout.html

An example of a custom JSON is shown here: https://www.jointjs.com/demos/kanban
